File: 6kyu/two_joggers.py
# determine how many laps need to be completed by both 'runners' to return to the same origin point, on different sized tracks,
# at the same time. the runners are running the exact same speed

def nbr_of_laps(x, y):
    
    # if the track sizes are the same, return 1 lap for each runner
    if x == y:
        return 1, 1
    
    else:
        # determine the max multiple
        product = x * y
        # determine each runner's quotient if running the max multiple distance
        max_x, max_y = product / x, product / y
        # determine the larger divisor of the two and subtract 1 as a starting point to decend from to find the least common multiple
        divisor = max([max_x, max_y]) - 1
        
        # loop while the divisor is greater than 1
        while divisor > 1:
            # if modulus returns 0 on both tracks, return the LCM
            if not max_x % divisor and not max_y % divisor:
                return max_x / divisor, max_y / divisor
            else:
                # reduce divisor and repeat
                divisor -= 1
        # return the initial quotient if no smaller common multiple is found
        return max_x, max_y

File: 6kyu/two joggers.c
#include <stdio.h>

typedef struct Laps_Pairing {
    int bob;
    int charles;
} laps;

laps nbr_of_laps(int x, int y) {
  
  
  if (x == y) {
    struct Laps_Pairing results;
    results.bob = 1;
    results.charles = 1;
    return results;
  }

  int product = x * y;
  int max_x = product / x;
  int max_y = product / y;
  int divisor;

  if (max_x > max_y) {
    divisor = max_x - 1;
  }
  else {
    divisor = max_y - 1;
  }
  
  while (divisor > 1) {
    if (max_x % divisor == 0 && max_y % divisor == 0) {
      struct Laps_Pairing results;
      results.bob = max_x / divisor;
      results.charles = max_y / divisor;
      return results;
    }
    else {
      divisor--;
    }
  }
  
  struct Laps_Pairing results;
  results.bob = max_x;
  results.charles = max_y;
  return results;
  
}
